skbio/metadata/_interval.py

```diff
@@ -0,0 +1,756 @@
+
+        Returns
+        -------
+        str
+            String representation of this ``Interval`` object.
+        '''
+        if self.dropped:
+            s = ('{}(dropped=True, locations={!r}, '
+                 'boundaries={!r}, metadata={!r})')
+            return s.format(self.__class__.__name__,
+                            self.locations, self.boundaries, self.metadata)
+        else:
+            s = ('{}(interval_metadata=<{!r}>, locations={!r}, '
+                 'boundaries={!r}, metadata={!r})')
+            return s.format(self.__class__.__name__,
+                            id(self._interval_metadata),
+                            self.locations, self.boundaries, self.metadata)
+
+    @experimental(as_of='0.5.0-dev')
+    def drop(self):
+        '''Drop this ``Interval`` object from the interval metadata it links to.
+
+        If the ``Interval`` object is dropped, you can still get values of
+        ``locations``, ``boundaries``, and ``metadata`` attributes, but you
+        can not change their values with the setters.
+
+        See Also
+        --------
+        skbio.metadata.IntervalMetadata.drop
+        '''
+        if not self.dropped:
+            self._interval_metadata.drop([self])
+
+    def _locations_boundaries_setter(self, locations=None, boundaries=None):
+        if self.dropped:
+            raise RuntimeError('Cannot change the dropped '
+                               'Interval object.')
+        # Casts to `list`, validation, sorting, and setting of `locations`
+        # and `boundaries` happen here.
+        if locations is not None:
+            # check iterability
+            try:
+                # check iterability
+                locations = list(locations)
+            except TypeError:
+                raise TypeError(
+                    'Cannotgive an iterable (%r) to `locations`.' % locations)
+
+            # check it is not empty
+            if not locations:
+                raise ValueError('Cannot give empty `location`.')
+            # check each contiguous span is in right format
+            for location in locations:
+                _assert_valid_location(location)
+
+            spans = len(locations)
+        else:
+            spans = len(self.locations)
+
+        if boundaries is not None:
+            try:
+                boundaries = list(boundaries)
+            except TypeError:
+                raise TypeError(
+                    'Cannot give a non-iterable (%r) '
+                    'to `boundaries`.' % boundaries)
+
+            if len(boundaries) != spans:
+                raise ValueError(
+                    'The length of boundaries must '
+                    'be equal to the length of locations {!r}.')
+
+            for boundary in boundaries:
+                _assert_valid_boundary(boundary)
+
+        if locations is None:
+            # `locations` and `boundaries` cannot both be omitted.
+            if boundaries is None:
```

Comment:
See the code comment I included. It's acceptable to use an assert because
it is asserting internal state that can't be reached by a user. This is
just a sanity check to make sure this private method isn't misused
elsewhere in the class by maintainers. It's okay if the check is disabled
at runtime. Does that make sense?

On Friday, September 2, 2016, Zech Xu <notifications@github.com> wrote:

> In skbio/metadata/_interval.py
> <https://github.com/biocore/scikit-bio/pull/1414#discussion_r77427217>:
>
> > +            except TypeError:
> > +                raise TypeError(
> > +                    'Cannot give a non-iterable (%r) '
> > +                    'to `boundaries`.' % boundaries)
> > +
> > +            if len(boundaries) != spans:
> > +                raise ValueError(
> > +                    'The length of boundaries must '
> > +                    'be equal to the length of locations {!r}.')
> > +
> > +            for boundary in boundaries:
> > +                _assert_valid_boundary(boundary)
> > +
> > +        if locations is None:
> > +            # `locations` and `boundaries` cannot both be omitted.
> > +            if boundaries is None:
>
> The reason I dont use assert it that the assert statements are removed in
> Python optimization mode. We want this check run all the time, right?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/1414/files/1837b1b85f84a93c77811f3c1a425f7da63c4a69#r77427217>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABwvwBbWYNFONgzVFJDEEkRWPgU4gXmWks5qmLz5gaJpZM4JoIbD>
> .
>


-- 
Sent from my phone.